docs/source/reference/oauth.md

```diff
@@ -0,0 +1,217 @@
+# JupyterHub and OAuth
+
+JupyterHub uses OAuth 2 internally as a mechanism for authenticating users.
+As such, JupyterHub itself always functions as an OAuth **provider**.
+More on what that means below.
+
+Additionally, JupyterHub is _often_ deployed with [oauthenticator](https://oauthenticator.readthedocs.io),
+where an external identity provider, such as GitHub or KeyCloak, is used to authenticate users.
+When this is the case, there are \*two nested
+
+This means that when you are using JupyterHub, there is always _at least one_ and often two layers of OAuth involved in a user logging in and accessing their server.
+
+Some relevant points:
+
+- Single-user servers _never_ need to communicate with or be aware of the upstream provider.
+  As far as they are concerned, only JupyterHub is an OAuth provider,
+  and how users authenticate with the Hub itself is irrelevant.
+- When talking to a single-user server,
+  there are ~always two tokens:
+  a token issued to the server itself to communicate with the Hub API,
+  and a second per-user token in the browser to represent the completed login process and authorized permissions.
+  More on this later.
+
+### Key OAuth terms
+
+- **provider** the entity responsible for managing.
+  JupyterHub is _always_ an oauth provider for JupyterHub's components.
+  When OAuthenticator is used, an external service, such as GitHub or KeyCloak, is also an oauth provider.
+- **client** An entity that requests OAuth tokens on a user's behalf.
+  JupyterHub _services_ or single-user _servers_ are OAuth clients of the JupyterHub _provider_.
+  When OAuthenticator is used, JupyterHub is itself also an OAuth _client_ for the external oauth _provider_, e.g. GitHub.
+- **browser** A user's web browser, which makes requests and stores things like cookies
+- **token** The secret value used to represent a user's authorization. This is the final product of the OAuth process.
+
+### The oauth flow
+
+OAuth flow is what we call the sequence of HTTP requests involved in authenticating a user and issuing a token, ultimately used for authorized access to a service or single-user server.
+
+It generally goes like this:
```

Comment:
Maybe restate that we have here 2 nested oauth flows.

Reply:
This is a general description of what one oauth flow is, so I don't think it makes sense to mention that there are two at this point in the doc. That comes later.

Reply:
Makes sense.